astropy/modeling/fitting.py

```diff
@@ -1181,81 +1280,81 @@ def __call__(self, model, x, y, z=None, weights=None,
-                                               self.fit_info['param_cov'])
-
-        model_copy.sync_constraints = True
-        return model_copy
 
-    @staticmethod
-    def _wrap_deriv(params, model, weights, x, y, z=None):
-        """
-        Wraps the method calculating the Jacobian of the function to account
-        for model constraints.
-        `scipy.optimize.leastsq` expects the function derivative to have the
-        above signature (parlist, (argtuple)). In order to accommodate model
-        constraints, instead of using p directly, we set the parameter list in
-        this function.
-        """
+class TRFLSQFitter(_NonLinearLSQFitter):
+    """
+    Trust Region Reflective (dogbox or Levenberg-Marqueardt) algorithm
+    and least squares statistic.
 
-        if weights is None:
-            weights = 1.0
+    Attributes
+    ----------
+    method : str
+        The method used to perform the fit (see notes)
+    fit_info :
+        A `scipy.optimize.OptimizeResult` class which contains all of
+        the most recent fit information
 
-        if any(model.fixed.values()) or any(model.tied.values()):
-            # update the parameters with the current values from the fitter
-            fitter_to_model_params(model, params)
-            if z is None:
-                full = np.array(model.fit_deriv(x, *model.parameters))
-                if not model.col_fit_deriv:
-                    full_deriv = np.ravel(weights) * full.T
-                else:
-                    full_deriv = np.ravel(weights) * full
-            else:
-                full = np.array([np.ravel(_) for _ in model.fit_deriv(x, y, *model.parameters)])
-                if not model.col_fit_deriv:
-                    full_deriv = np.ravel(weights) * full.T
-                else:
-                    full_deriv = np.ravel(weights) * full
+    Notes
+    -----
+    The ``method`` determines which fitter algorithm to use:
+        ‘trf’ : Trust Region Reflective algorithm, particularly suitable
+            for large sparse problems with bounds. Generally robust method.
+        ‘dogbox’ : dogleg algorithm with rectangular trust regions, typical
+            use case is small problems with bounds. Not recommended for
+            problems with rank-deficient Jacobian.
+        ‘lm’ : Levenberg-Marquardt algorithm as implemented in MINPACK.
+            Doesn’t handle bounds and sparse Jacobians. Usually the most
+            efficient method for small unconstrained problems.
+    """
 
-            pars = [getattr(model, name) for name in model.param_names]
-            fixed = [par.fixed for par in pars]
-            tied = [par.tied for par in pars]
-            tied = list(np.where([par.tied is not False for par in pars],
-                                 True, tied))
-            fix_and_tie = np.logical_or(fixed, tied)
-            ind = np.logical_not(fix_and_tie)
+    def __init__(self, calc_uncertainties=False, method='trf', check_bounds=False):
```

Comment:
Unless I'm missing something, it doesn't seem necessary for `method` to be in the constructor. If passed to `_run_fitter`, then the same instance of a fitter can be used with different methods.

Reply:
Sure one can do this, but it then requires a small interface change to `LevMarLSQFitter`, namely enabling one to pass a dummy `method` argument when calling the fitter.

Reply:
I have been thinking more about this problem. Having a selectable fitter method/algorithm is not how any of the other fitters operate in fitting. Maybe it is better to introduce a separate fitter class for each fitting option.

Reply:
See 05b073a for the implementation described above.